### DISCUSSION TITLE
Next.js + Redis + Kubernetes

### DISCUSSION THREAD

**Discussion by ezeparziale** [vercel/next.js#58703] [Help]
### Summary

I'm trying to configure a Next.js web app with 2 pods sharing a cache on Redis.

I'm following these examples:

- Redis config: [https://github.com/vercel/next.js/tree/canary/examples/cache-handler-redis](https://github.com/vercel/next.js/tree/canary/examples/cache-handler-redis)

- Docker config: [https://github.com/vercel/next.js/tree/canary/examples/with-docker](https://github.com/vercel/next.js/tree/canary/examples/with-docker)

- Next.js doc: [https://nextjs.org/docs/app/building-your-application/deploying#self-hosting](https://nextjs.org/docs/app/building-your-application/deploying#self-hosting)

I installed `sharp` for image optimization without any additional configuration.

My `next.config.js` is:

```node
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    incrementalCacheHandlerPath:
      process.env.NODE_ENV === "production"
        ? require.resolve("./cache-handler.js")
        : undefined,
  },
  output: "standalone",
};

module.exports = nextConfig;
```

I have few questions:
1. Do I need to add: `isrMemoryCacheSize: 0, // disable default in-memory caching` in `next.config.js` ?
2. What happens with image optimization between pods? Do the pods need to share some volume?
3. Is there any other configuration needed?

I have my example in: [https://github.com/ezeparziale/nextjs-k8s](https://github.com/ezeparziale/nextjs-k8s)

### Example

https://github.com/ezeparziale/nextjs-k8s

**Comment by better-salmon** [comment]
Hey @ezeparziale! I can speak for the [cache-handler-redis](https://github.com/vercel/next.js/tree/canary/examples/cache-handler-redis) example and the `@neshca/cache-handler` package.

First of all, please update the `@neshca/cache-handler` package to the latest version, 0.6.1, at the time of writing, as the documentation is applicable only for this version. You can use [this example](https://caching-tools.github.io/next-shared-cache/redis-stack).

> Do I need to add: isrMemoryCacheSize: 0, // disable default in-memory caching in next.config.js?

The `IncrementalCache` class from the `@neshca/cache-handler` ignores the `isrMemoryCacheSize` option. However, you can specify whether to use in-memory cache by adding the Handler from `@neshca/cache-handler/local-lru` to the return of `OnCreationHook`. **So, in your case, you don't need to define this option.**

The `@neshca/cache-handler/local-lru` is considered a **fallback** if the remote cache store fails.

> What happens with image optimization between pods?

The images are not stored in the cache, and the image processing happens locally on each pod. In my opinion, you should consider using on-the-fly image processing and a CDN, e.g., imgproxy.

> Do the pods need to share some volume?

You may use shared volumes for your pods in combination with a custom cache handler. However, it's worth the effort only if you are using the Pages directory and want a 0% chance of showing a stale page to the user. You can find more information in [this docs section](https://caching-tools.github.io/next-shared-cache/configuration/use-file-system) which reflects the usage of the file system.

If you are using only the App directory, the best option is to disable the file system, enable TTL for both Redis and LRU handlers. I have a guide about TTL [here](https://caching-tools.github.io/next-shared-cache/configuration/ttl).

> Is there any other configuration needed?

- Using Redis as a cache store always implies configuration of the Redis server itself, which you did in your example.
- Consider configuring image processing and a CDN service (Optional).
- You always can opt-out of the pre-configured cache Handler and implement all methods yourself for extra flexibility (Optional).

---

I tried to answer your questions as thoroughly as possible. If you have any other questions, please feel free to ask. I'll be happy to help you.

And if the maintainers of Next.js have anything to add, it would be great.

**Reply by ezeparziale** [reply]
Wow!!! amazing reply!!!

I am reading the new documentation and I upgrading my project example.

> [!NOTE]
> I am using Next.js 14 and App router

I wiil use this:

```javascript
IncrementalCache.onCreation(async () => {
    let redisCache, localCache;

    const useTtl = true;

    if (process.env.REDIS_AVAILABLE) {
        await client.connect();
 
        redisCache = await createRedisCache({
            client,
            useTtl,
        });
     
        localCache = createLruCache({
            useTtl,
        });
    }

    const cache = [redisCache, localCache];
    
    return {
        cache,
        useFileSystem: false
    };
});
```
